src/reanimated2/animation/transformationMatrix/matrixUtils.tsx

```diff
@@ -0,0 +1,433 @@
+  'worklet';
+  const cos = Math.cos(angle);
+  const sin = Math.sin(angle);
+  switch (axis) {
+    case 'z':
+      return [
+        [cos, sin, 0, 0],
+        [-sin, cos, 0, 0],
+        [0, 0, 1, 0],
+        [0, 0, 0, 1],
+      ];
+    case 'y':
+      return [
+        [cos, 0, -sin, 0],
+        [0, 1, 0, 0],
+        [sin, 0, cos, 0],
+        [0, 0, 0, 1],
+      ];
+    case 'x':
+      return [
+        [1, 0, 0, 0],
+        [0, cos, sin, 0],
+        [0, -sin, cos, 0],
+        [0, 0, 0, 1],
+      ];
+  }
+}
+
+function norm(x: number, y: number, z: number) {
+  'worklet';
+  return Math.sqrt(x * x + y * y + z * z);
+}
+
+function transposeMatrix(matrix: AffiniteMatrix): AffiniteMatrix {
+  'worklet';
+  const m = flatten(matrix);
+  return [
+    [m[0], m[4], m[8], m[12]],
+    [m[1], m[5], m[9], m[13]],
+    [m[2], m[6], m[10], m[14]],
+    [m[3], m[7], m[11], m[15]],
+  ];
+}
+
+function innerProduct(a: number[], b: number[]) {
+  'worklet';
+  if (a.length !== b.length) {
+    console.warn(
+      `Cannot calculate inner product of two vectors of different length. Length of ${a} is ${a.length} and length of ${b} is ${b.length}.)`
+    );
+    return 0;
+  }
+  return a.reduce((acc, _, i) => acc + a[i] * b[i], 0);
+}
+
+function projection(u: number[], a: number[]) {
+  'worklet';
+  const scaler = innerProduct(u, a) / innerProduct(u, u);
+  return u.map((e) => e * scaler);
+}
+
+function subtract(a: number[], b: number[]) {
+  'worklet';
+  return a.map((_, i) => a[i] - b[i]);
+}
+
+function scale(u: number[], a: number) {
+  'worklet';
+  return u.map((e) => e * a);
+}
+
+function gramSchmidtAlghoritm(matrix: AffiniteMatrix): {
+  rotationMatrix: AffiniteMatrix;
+  skewMatrix: AffiniteMatrix;
+} {
+  'worklet';
+  const [a0, a1, a2, a3] = matrix;
+
+  const u0 = a0;
+  const u1 = subtract(a1, projection(u0, a1));
+  const u2 = subtract(subtract(a2, projection(u0, a2)), projection(u1, a2));
+  const u3 = subtract(
+    subtract(subtract(a3, projection(u0, a3)), projection(u1, a3)),
+    projection(u2, a3)
+  );
+
+  const [e0, e1, e2, e3] = [u0, u1, u2, u3].map((u) =>
+    scale(u, 1 / Math.sqrt(innerProduct(u, u)))
+  );
+
+  const rotationMatrix: AffiniteMatrix = [
+    [e0[0], e1[0], e2[0], e3[0]],
+    [e0[1], e1[1], e2[1], e3[1]],
+    [e0[2], e1[2], e2[2], e3[2]],
+    [e0[3], e1[3], e2[3], e3[3]],
+  ];
+
+  const skewMatrix: AffiniteMatrix = [
+    [
+      innerProduct(e0, a0),
+      innerProduct(e0, a1),
+      innerProduct(e0, a2),
+      innerProduct(e0, a3),
+    ],
+    [0, innerProduct(e1, a1), innerProduct(e1, a2), innerProduct(e1, a3)],
+    [0, 0, innerProduct(e2, a2), innerProduct(e2, a3)],
+    [0, 0, 0, innerProduct(e3, a3)],
+  ];
+  return { rotationMatrix, skewMatrix };
+}
+
+export function decomposeMatrix(
+  unknownTypeMatrix: AffiniteMatrixFlat | AffiniteMatrix
+): TransformMatrixDecomposition {
+  'worklet';
+
+  const matrix: AffiniteMatrixFlat = maybeFlattenMatrix(unknownTypeMatrix);
```

Comment:
I believe this should work without `: AffiniteMatrixFlat`, the type should be inferred from the return type of `maybeFlattenMatrix`.